### PR TITLE
Refactor joiner functionality

### PR DESCRIPTION
**Pull request by r-b-g-b**

### CODE REVIEW

**Review by Ethan-bradley — COMMENTED** [review]
This code is a revised version of the former joiner.py. It combines the air quality and gps files and then in the second commit file sends them to the database. This code creates an air quality file that almost exactly matches the original, the differences being a minor problem involving one data point.

**Comment by r-b-g-b** [comment]
Thanks Ethan, that's a good description of what this is about. I won't make a habit of comment-less PRs, but wanted to put the code somewhere so we could look at it at the meetup tonight.

The discrepancy Ethan refers to is a slight difference in the algorithm to link GPS and air quality measurements. The original only looked for GPS points after the air quality measurement, this one will take the closest point even if it earlier in time.

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py:16]: You are using the [NumPy docstring style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy). The [Google docstring style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html). Whichever you pick, just make sure folks stick with it throughout the codebase.

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py]: Looking higher level, how is this bit of code supposed to be used? I assume it will be called shortly after files are uploaded. If that's the case, you don't need a main function as this isn't going to be a script called from the command line. Similarly, you aren't going to accept file paths as an input.

Consider updating `join()` to accept the raw files rather than their paths. This will allow you to call `join()` from your code responsible for handling the uploads.

If you have a need to call this code from the command line, create a [management command](https://docs.djangoproject.com/en/2.1/howto/custom-management-commands/).

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py]: This definitely needs testing. At a minimum, you should pass in data and ensure the joined data is written to the database as expected.

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py:228]: s/numeric/float

\* `s/` is the command for search and replace in vi and other Unix tools: http://www.linfo.org/vi/search.html.

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py:257]: This is never called. Dead code should not be in the code base. Remove it, and restore it when it is actually in use.

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py]: Use [`create()`](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#create) to instantiate and save the object to the database in one call:

```python
models.Data.objects.create(
  session_data=session_data,
  value=row['measurement'],
  time=row['time'],
  latlon=geos.Point(row['lon'], row['lat'])
)
```

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py:266]: Use [`bulk_create()`](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#bulk-create) to create all items in a single database call:

```python
models.Data.objects.bulk_create([
  models.Data(session_data=session_data,
    value=row['measurement'],
    time=row['time'],
    latlon=geos.Point(row['lon'], row['lat']))
  for _, row in joined_data.iterrows()
])
```

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/dustrak.py]: This definitely fits on a single line: `message = f"{n_dropped} air quality samples dropped that were not within {tolerance} seconds of a GPS sample."`.

**Review by clintonb — CHANGES_REQUESTED** [review]
I noticed a couple minor things:
The PR title is still prefixed with "WIP". Remember to remove this before merging.

Let me know if you have questions or need assistance with testing.

File: woeip/apps/air_quality/dustrak.py
Comment (clintonb) [reply woeip/apps/air_quality/dustrak.py]: This is an FYI. Use `bulk_create()` for this function, assuming the function is not removed.

File: woeip/apps/air_quality/dustrak.py
Comment (r-b-g-b) [reply woeip/apps/air_quality/dustrak.py:16]: Will stick with numpy style unless you think Google docstrings are better!

Comment (clintonb) [reply woeip/apps/air_quality/dustrak.py:16]: I don't think either is better than the other. Use whichever you all prefer.

File: woeip/apps/air_quality/dustrak.py
Comment (r-b-g-b) [reply woeip/apps/air_quality/dustrak.py:257]: This will be useful when we create the data upload forms. For now there's at least a test that uses this.

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/tests/test_dustrak.py]: Disable pylint warnings for the line with the issue rather than the entire file.

File: woeip/apps/air_quality/dustrak.py
Comment (TangoYankee) [reply woeip/apps/air_quality/dustrak.py:16]: Is it a style preference? Or is it common to apply logic that depends on the format of the docstring?

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/tests/test_dustrak.py]: This is not needed. Factories generate saved objects.

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (clintonb) [inline woeip/apps/air_quality/tests/test_dustrak.py]: You should validate the values saved to the DB.

**Review by clintonb — COMMENTED** [review]
No need to await my approval.

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (TangoYankee) [reply woeip/apps/air_quality/tests/test_dustrak.py]: This is suppressing warnings for use of the ```target_data``` name. For lines 25 and 44, it makes sense to suppress the warning; pylint fails to recognize they are function calls. However, I am less sure about lines 17-22. There, ```target_data``` is defined as a variable within the ```target_data``` function. Would it be better to rename ```target_data``` in lines 17-22 to ```target_output```, rather than simply suppress the warning? Or does pytest.fixtures rely on the logic of calling the function within itself?

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (TangoYankee) [reply woeip/apps/air_quality/tests/test_dustrak.py]: As in, lines 62-65 of ```tests/factories.py``` save the data, when they are created using ```factory```? Consequently, just delete line 47 of ```tests/test_dustrak.py``` ?

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (TangoYankee) [reply woeip/apps/air_quality/tests/test_dustrak.py]: Rather than simply checking the lengths, crack the files open with a for-loop and assert that each line of ```models.Data.objects.all()``` matches ```target_data``` ?

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (clintonb) [reply woeip/apps/air_quality/tests/test_dustrak.py]: The call to the base class saves the object, not lines 62-65.

Yes, delete this line.

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (clintonb) [reply woeip/apps/air_quality/tests/test_dustrak.py]: The fix here is to not use a fixture. This data is only needed by the one test, so there is no need to make a fixture (which you might share with multiple test modules). Instead, just call the function from your tests.

If you must disable a pylint warning, use the name, not the code (e.g. `redefined-outer-name`) to make it clear what's going on without an extra lookup.

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (TangoYankee) [reply woeip/apps/air_quality/tests/test_dustrak.py]: Edit: used a numpy.allclose, similar to the one used in the joiner test_joiner. Limited it to a check of measurements in target data against values in database storage.

File: woeip/apps/air_quality/tests/test_dustrak.py
Comment (TangoYankee) [reply woeip/apps/air_quality/tests/test_dustrak.py]: It took me forever to find a [good list of error names](https://github.com/janjur/readable-pylint-messages/blob/master/README.md). Do you just ...*know* them?

Comment (clintonb) [reply woeip/apps/air_quality/tests/test_dustrak.py]: Nope. When you run pylint the errors are printed in parentheses.

Comment (TangoYankee) [reply woeip/apps/air_quality/tests/test_dustrak.py]: Hmm, all I saw were the messages and not the names. Guess I know what to look for next time.

**Review by TangoYankee — APPROVED** [review]

**Pull request merged**
